Design experiment

victlu opened the pull request:
A PR for discussion.  Not intended to be merged.

Currently, there are 3 different designs…

./examples3 – An interpretation for a simpler API/SDK (from Victor)
./experimental – An interpretation from a .NET perspective (From Noah)
./examples2 – Trying to implement an interpretation of existing OTel Metric Spec (From Victor)

victlu committed:
Example of a new API/SDK

victlu committed:
API/SDK using abstract class

noahfalk committed:
Merge pull request #2 from victlu/new-api-sdk

New api sdk

victlu committed:
Setup more concurrent counters

victlu committed:
experimenting...

victlu commented on experimental/ConsoleApp6/ConsoleApp6/Api.cs:
```diff
@@ -0,0 +1,118 @@
+        public LabelSet DefaultLabels { get; }
+
+        public void Record(int measurement) { }
+        public void Record(int measurement, string labelName1, string labelValue1) { }
```
Looks like you're separating Key and Value.  So, why not Record(..., string[] keys, string[] values)?

noahfalk replied:
Sorry I completely missed some of these comments. My bad!
A few reasons:
a) It is one less allocation per metric recording (cost ~10ns)
b) passing strings at Record implies the names can be different every time which means the SDK has to search and/or sort them

victlu replied:
RE: b.  Although I agree, I don't think that's your top priority concern.

victlu commented on experimental/ConsoleApp6/ConsoleApp6/Api.cs:
```diff
@@ -0,0 +1,118 @@
+
+    //TODO: we could derive from Metric but that would expose the Record() API. Does that matter?
+    //TODO: If we don't derive from Metric perhaps we should replicate the other properties?
+    public struct Counter
```
I noticed you have a struct for Counter as a wrapper for a Metric class.  Is this faster?

noahfalk replied:
It was a silly way to supress the APIs that would have shown up if I had inherited. What I should have done (and at this point have done) is mark the APIs protected.

victlu commented on experimental/ConsoleApp6/ConsoleApp6/Api.cs:
```diff
@@ -0,0 +1,118 @@
+
+
+
+    public static class MetricAggregations
```
Wondering if any Aggregations should be part of the API.  If it is, how would new aggregations be added?

noahfalk replied:
These are just convenient singletons to reduce allocations. The intent is that anyone can derive new types and invoke new to create an instance of their new type.

victlu commented on experimental/ConsoleApp6/ConsoleApp6/Examples.cs:
```diff
@@ -0,0 +1,111 @@
+{
+    class SquidAnalyzer
+    {
+        static readonly MetricSource metrics = new MetricSource("SquidLibrary",
```
Need to discuss if we really want to have a "Provider" as the starting point.

I think it makes more sense to just "new Counter(...)" immediately.  And employ late binding to attach to a "provider".

victlu commented on experimental/ConsoleApp6/ConsoleApp6/Examples.cs:
```diff
@@ -0,0 +1,111 @@
+                { "SquidBetaBuild", "true" },
+                { "SeafoodFeatureLevel", "4" }
+            });
+        static readonly Counter squidsAnalyzedCounter = metrics.CreateCounter("squids-analyzed");
```
I'm wondering if there's an easy way (preferably at compile time) to get the caller's fully qualified class name.  That can act as the "default" name for this counter.  It may be done with reflections (but it's slow).

victlu replied:
Just like ILogger<ClassCategory>, we can also use generics to take a class name that we use for library name and we can get library version via reflection.

victlu commented on experimental/ConsoleApp6/ConsoleApp6/Examples.cs:
```diff
@@ -0,0 +1,111 @@
+            string labelValue2 = "shrimp";
+            string labelValue3 = "clam";
+            string labelValue4 = "soup";
+            squidsAnalyzedCounter.Add(1,
```
This looks pretty awkward to me in passing in key/values...

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+
+Micrometer (MM) - https://micrometer.io/docs/concepts
+
+1. Do the names that OpenTelemetry chose match industry conventions? 
```
I'm wondering maybe we leave the "naming" convention match to the SDK.  Thus, API just needs to collect all the component fields, and let the SDK do what they want with forming a correct "name" for the exporter.

Basically, I'm saying maybe we are not so concern with naming from API side as it can be adjusted by the SDK.

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+Micrometer (MM) - https://micrometer.io/docs/concepts
+
+1. Do the names that OpenTelemetry chose match industry conventions? 
+For the MM library there are different types of Meters such as timers, counters, gauges, etc. Depending on their type, meters implement operations such
```
I think I heard may calls for a "timing" type counter.  Maybe we include a TimeRange counter of the form...

using (var timer = new Timer(...))
{
    // Do Operations...
}

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+As best I can tell the best conceptual mapping is that MM's Meter == OT's bound instrument. It doesn't worry me too much that OT would use a different
+name for the same concept, what worries me more is if OT is reusing the same names but altering the concept they apply to.
+
+2. MM has no equivalent abstraction for OT's Meter or Instrument. It goes straight from MeterProvider (MeterRegistry in MM-terminology) to a bound-instrument 
```
Agree.  I would propose we employ "late binding" to attach a Provider.

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+If MM has been succesful with these simpler abstractions it raises the question what value is OT getting from its more complex 4 layers of
+abstraction (MeterProvider->Meter->Instrument->BoundInstrument)?
+
+3. MM landed on a different set of instrument types than OT did. It would be good to understand why we believe OT's instrument set is the right set to pick?
```
Yeah, I think the OT SIG also pointed this out.  I think Reiley will likely drive this to some better conclusion.

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+or is there a scenario where the atomicity of the measurements matter?
+
+6. MM has a lot of API surface dedicated to configuring Meters (in MM terminology) and tags. I am guessing this is the kind of functionality that would be sitting in OT's
+Views API?
```
This will be a HUGE topic for SDK side.  And probably will be very vendor opinionated in those discussions!

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+Histogram     ValueRecorder
+NA            *Observer
+
+4. Prometheus uses Observe() API where OT uses Record() API. Prometheus doesn't appear to have asynchronous/callback style APIs.
```
I think we do need an "async" callback for reporting data.  Or at least some examples of how to do this.  Need to resolve the issue of Sum vs Incremental dataset.  Cause any "Observed" value implies a Time Range, and thus, some kind of aggregation.  This needs to be reviewed a bit deeper I think.

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+var counter = new CounterOptions { Name = "my_counter" } // this defines the metric
+metrics.Measure.Counter.Increment(counter); // increments the value of the metric stored in this IMetrics
+
+3. Metrics are never bound. Tags can be provided at the time a measurement is captured and merged with tags set via global tags in the IMetrics interface
```
I'm wondering... Maybe we setup a memory storage for binding a set of KVs.  Then, we just pass the id of that entry, instead of "binding" a meter.  This way, any "bound" KVs can be shared by any meter and addressible with an id.

victlu commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+Double and long. The typical distinction between counters, gauges, histograms, etc are defined separately as aggregations using the View API.
+3. Measures are created directly from a static factory method and they are not associated with any particular collection pipeline. A separate StatsRecorder object provides
+the data collection API.
+4. The StatsRecorder collection API builds up a list of measurements first, then commits it with a single API call. I didn't see any simpler option presented in the samples
```
We likely need a BatchRecord type functionality.  To allow our user to inject data for the same "timeframe".  I don't know what other scenarios requires "atomicity" (outside of marking same timeframe) but would like to find out more from others what the need here is.

victlu committed:
Introduce LabelSet

victlu committed:
Try wrapping int/double into MetricValue struct

reyang commented on example3/API/MetricSource.cs:
```diff
@@ -0,0 +1,33 @@
+        /// <summary>
+        /// Let SDK know when new counters are being created
+        /// </summary>
+        public abstract bool OnCreate(MetricBase counter, LabelSet labels);
```
Would this assume the SDK is injected before the counters are being created?

victlu replied:
I would prefer to allow the SDK to attach whenever.  But you make a good point, about what we do with these OnCreate events if an SDK is currently not attached.  Or if we do anything after a SDK is attached for previously created item.

victlu replied:
This issue is no longer applicable with current implementation of MetricSource

victlu committed:
Add Batch Record example.  Add LabelHistogram example

victlu committed:
Refactor based on discussion with Noah

victlu committed:
Create a MetricProvider as a top level.

victlu committed:
Refactoring...

victlu committed:
Churning...

victlu committed:
Add Hints to Meter

victlu committed:
Introduce a Queue for buffering incoming recordings

victlu committed:
Use Span<> for MetricValue

victlu committed:
Added Benchmark tests.

noahfalk committed:
Adding more design Q&A

victlu committed:
Example for Async callbacks

reyang commented on experimental/design-notes.txt:
```diff
@@ -0,0 +1,112 @@
+For the OT spec Meters are factories that create instruments and it is the instruments that have the data collecting APIs such as Add() or Record(). Instruments
+aren't associated with a specific set of tags (labels) unless they are first bound.
+As best I can tell the best conceptual mapping is that MM's Meter == OT's bound instrument. It doesn't worry me too much that OT would use a different
+name for the same concept, what worries me more is if OT is reusing the same names but altering the concept they apply to.
```
Since Micrometer is Java specific, there might be less concern.
We can check what's the convention in Prometheus as it is well established and it is cross-langages.

victlu replied:
added "eamples-other-apis.md" document to show examples of other API patterns.

victlu committed:
add code examples from other apis

victlu committed:
Benchmark for Field vs Generic

victlu committed:
Use generics for passing metric value

victlu committed:
Use generics for passing metric value

victlu committed:
Rename MetricProvider to MetricSource

victlu committed:
Add benchmarks for LabelSet

victlu committed:
Add GroceryStore example

victlu committed:
Fix minor bug

victlu committed:
Add a way to configure Label dimensions for aggregation

victlu committed:
Update output of Grocery example

victlu committed:
Add Exporters

noahfalk committed:
More design discussion

victlu committed:
Refactor out DotNet runtime portion

victlu committed:
Refactor into csproj and sln files

victlu committed:
Adding potentially controversial principle statement.

victlu committed:
Refactor for SumCountMinMax and Value Tuple

victlu committed:
Add benchmark

victlu committed:
Refactor to allow configuration of Aggregators

victlu committed:
Try a RateCounter from .NET portion

noahfalk committed:
Q&A cleanup

separated API questions from SDK questions
moved a few items to a graveyard section
reframed class vs. interface topic as question with specific scope and options
added proposed answer and rationale to class vs. interface topic

noahfalk committed:
commenting on the principles

victlu commented on example3/readme.md:
```diff
@@ -0,0 +1,580 @@
+
+[Noah]: I agree with the Event sourcing principle but I would propose MetricBase is
+the implementation of the source concept in this pattern. So the data flow is:
+MetricBase -> Listener instead of MetricBase -> Source -> Listener.
```
Current code has MetricBase -> Listener which implements the EventSource/EventListener pattern.  This WHOLE pair is considered a "source" (maybe I should renamed this to be Provider) which the OTel sdk attaches to for the pipeline building.

victlu commented on example3/readme.md:
```diff
@@ -0,0 +1,580 @@
+- OTel may implement and deliver counters (derived from .NET MetricBase) with it's
+own Semantic. (i.e. UpCounter with Inc(), ValueRecorder with Record(), etc...)
+
+[Noah]: This principle worries me. I want .NET developers to treat .NET metrics as being 
+a single unified concept. Ie if I am writing a .NET library I want to intrument it:
+a) Without having any reference to an OT specific library
+b) Without mentally apportioning some metrics as being intended for OT usage and others
+are not intended for OT usage.
```
The unifying concept needs to be the MetricBase.  

I think .NET will provide the basic set of "instruments" (i.e. A Sum and a Gauge) that will cover 95+ % of basic needs.  However, some semantics are opinionated, and people can take OTel API if they want those opinionated "instrumnents" (i.e. UpDownCounter, Summary, Histogram, etc...).  Plus, people and other 3P libraries will derived their own from MetricBase with it's own semantics.

victlu commented on example3/Microsoft.Diagnostics.Metric/MetricLabel.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <summary>
+        /// Return Array of Tuple&lt;Key,Value&gt;.
+        /// </summary>
+        public virtual (string name, string value)[] GetLabels()
```
@noahfalk Any concerns with use of Value Tuple?  I think we can use array of Value Tuple for our "LabelSets" that is a good compromise between usability vs performance.

noahfalk replied:
For API usability .NET team typically recommends using a named struct with properties. It will have the same performance characteristics as the tuple.

victlu committed:
Use MetricLabel in SDK

victlu committed:
update readme after talking to Noah

victlu committed:
Add local-agg type .NET Counters

victlu committed:
Clarify naming of class

victlu committed:
Experiment with LastValueGauge

victlu committed:
Move Hints to MetricBase

victlu committed:
Tinkering...

victlu committed:
Tinkering

victlu committed:
Refactor Export pipeline

victlu committed:
Try using Protobuf for Exporter

victlu committed:
Encode/Decode ExportMetricsServiceRequest

noahfalk committed:
Lots of changes

- create meters with 'new', not a factory
- eliminated MetricSource, Listener subscribes to Meters directly (ie Meter is source Source)
- eliminated all the SDK meter types. Meters could be extendable by the SDK but none of our basic scenarios should require it
- added observable exporters
- added bound meters and cached listener cookies so the listener can cache the transformation from labelset to aggregator List
- on the API side split labels into names and values. Names are specified at construction time, values at recording time
- broke some of the SDK configuration API, the label histogram, and the OLTP exporter - it could be restored but haven't gotten to it yet

victlu committed:
Turn on exporter

victlu committed:
Get OTLPExporter working

victlu committed:
Add LibraryName and LibraryVersion

victlu committed:
Add more examples

victlu committed:
OTel API wrapper around .NET API

victlu committed:
Try out new Metric protocol

victlu committed:
Add Benchmarks for metric proto

victlu committed:
Update Bench results

victlu committed:
tinkering with OTel API

victlu committed:
Add DI example unittest

noahfalk commented on example3/Example/Library.cs:
```diff
@@ -0,0 +1,85 @@
+    public class Library
+    {
+
+        static Counter s_c1 = new Counter("MyLibrary", "1.0.0", "c1", new string[] { "Label1", "Label2" });
```
At the moment I'm a fan of the idea that counters are primarily distinguished by their name and that if library_name or version exist at all, they are optional. A few reasons:
1. It is simple
2. It aligns with what other APIs already do (Prometheus, Micrometer, ILogger)
3. I've never seen particularly good guidance from OpenTelemetry about how developers are expected to set these fields (in particular version) and what value they are getting in return for adding this extra complexity to their code.

reyang replied:
In tomorrow's Metrics API/SDK SIG meeting I'll bring the version question.
My understanding is that [OpenTelemetry tracer version](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#get-a-tracer) is used to describe the instrumentation library version, which is controlled by the semantic version of the semantic convention (I know it is very mouthful).

reyang replied:
And in metrics we (I mean OpenTelemetry community) have intention to keep the metrics provider API consistent with tracer provider API.

noahfalk replied:
I'm not opposed to the data existing, put I am questioning making it a required constructor parameter in the .NET API. For example these values are not required inputs to LoggerFactory creation so there has to be some path where OT can consume telemetry when the developer hasn't provided this metadata.

reyang replied:
I agree, I don't think putting a version string in the API is a common practice in .NET. I've seen it in some Java IoC (Inversion of Control) framework.
I think OpenTelemetry MeterProvider will have the version info, just to keep it consistent with TracerProvider API.
For .NET, I think we don't add it in the BCL libraries, we use convention instead (similar like `ActivitySource("MyCompany.MyService.MyComponent")`).

victlu replied:
Version is an optional component per current Spec.  Also, we will want to clarify if version plays any part in the "identity" of a metric.

victlu replied:
@noahfalk, you took out library name/ver in new Counter().  However, we need to find a way to make that available as OTel currently requires it.

noahfalk replied:
It is inside MetricSource now

noahfalk replied:
(and MetricSource is optional constructor parameter)

noahfalk commented on example3/OpenTelemetry.Metric.Api/Meter.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    public class Counter1D
```
These weren't in the spec, were you proposing they needed to be created?

reyang replied:
I was curious about this, too. Was just trying to figure out if this was added by Noah or Victor (as this is a big PR). 🤣

reyang replied:
Hey, before this, quick question - should I comment on this giant PR?

victlu replied:
I added these as WRAPPER classes to allow compile time check on the # of dim value a user is allowed.  If we did not have this, it is up to the user to match the # of dimension values with the same # when they created it.

Also, a larger item we need to discuss is the flexibility and freedom of adding arbitrary # of labels.  The .NET API currently requires us to declare the # of dim names upfront.  This maybe conflicting with current metric spec.

cijothomas replied:
many existing systems (applicationinsights, prometheus-dot-net, msft internal metric system) require dimension keys to be given in advance. Otel doesn't require that -you can add any key/value at any time.

cijothomas replied:
@victlu Does knowing ahead the names of dimension keys allow a perf optimization?

noahfalk replied:
> The .NET API currently requires us to declare the # of dim names upfront. This maybe conflicting with current metric spec.

I did this as an experiment because I saw Prometheus had done it and the API seemed nice to me. It certainly conflicts with the current spec : )

> @victlu Does knowing ahead the names of dimension keys allow a perf optimization?

Yeah several:
 - don't need to check the key names against the key names in any other label set when determining equality within a particular meter's time series
 - don't need to sort because there is a canonical ordering of the dimension names, 
 - don't need to store the key names as part of a time series label set (lower memory use)

cijothomas replied:
Thanks for confirming. I guess these advantages are the reason which motivated all the existing libraries to require label keys in advance.

cijothomas replied:
From Otel Metric community meeting, the main reason for *not* requiring labelkeys in advance was to avoid requiring users to always provide the label values,, even when label value is not available.
For example:
```
var ctr  = new Counter("countername", "label1","label2");
ctr.Record(100, "label1Value"); // this will throw as label2 is missing.
```

Additionally, one of the goal is to have Baggage or Resource becoming additional labels.
Baggage.Current.Add("user","cijo")
ctr.Record(100, "label1Value", "label2Value"); // this will internally can pick additional labels from the Baggage. (assume SDK is configured to do so).

victlu replied:
I'm actually challenging pre-defined dim names.  I think we should take a collection of labelset.  A labelset is a collection of key/value pairs.  This requires "merging" of multiple lableset into an effective labelset.  But allows immutability for labelset that is provided (i.e. baggage and attributes, bound instruments, etc...) to be cached.

noahfalk replied:
> From Otel Metric community meeting, the main reason for not requiring labelkeys in advance was to avoid requiring users to always provide the label values,, even when label value is not available.

How about?
````
ctr.Record(100, "label1Value", null)
````

Fwiw I'm still not convinced predefined names are the way to go either, but slotting in null doesn't seem that bad? Alternately if the labels at the end of the list were missing that doesn't have to throw an error, the API could just assume that anything missing at the end is automatically null.

> Additionally, one of the goal is to have Baggage or Resource becoming additional labels.

I don't think this conflicts. The label names define the names of labels that the instrumentation code is providing. There is no problem with the SDK adding additional name/value pairs.

> I'm actually challenging pre-defined dim names. I think we should take a collection of labelset.

There is a static labels property which is intended to hold labels where both the name and value are known when the counter is declared. The list of label names are for labels where the value is unknown until the point in time where the measurement is made, it would be impossible to include such a label into a labelset provided to the constructor.

noahfalk reviewed (changes requested):
Just making sure this doesn't get accidentally get merged because I want main matching upstream. It isn't anything about my sentiment on the change ; )

victlu committed:
Tinkering

victlu committed:
Simple DI example

cijothomas commented on example3/OpenTelemetry.Metric.Api/Meter.cs:
```diff
@@ -0,0 +1,173 @@
+        Counter CreateCounter(string name, Dictionary<string,string> labels);
+        Counter CreateCounter(string name, Dictionary<string,string> labels, params string[] dimNames);
+
+        Counter1D CreateCounter(string name, string dn1);
```
From a initial glance at the PR, this looks like a user needs to take a dependency on Otel.Api to start emitting metric? Or can they do it with pure .NET (i.e by depending on Ms.Diagnostic.Metric alone? which I would assume will become System.Diag.DiagnosticSource.)

noahfalk replied:
My current stance is that I am aiming to have the vast majority of library instrumentation scenarios (95+%) are doable with .NET API alone. I'm not sure yet what @victlu had in mind with these Counter#D types : )

victlu replied:
The Counter#D types was to limit # of dim values taken in Add() calls.  Otherwise, it's an open string array with possible mismatch in # of values provided.

cijothomas commented on example3/OpenTelemetry.Metric.Sdk/Aggregators/SumCountMinMax.cs:
```diff
@@ -0,0 +1,42 @@
+using Microsoft.Diagnostics.Metric;
+using OpenTelemetry.Metric.Sdk;
+
+namespace OpenTelemetry.Metric.Sdk
```
I only see aggregators in Otel SDK. Won't .NET come with a set of builtin aggregators for common things like MMSC?

noahfalk replied:
I'm interested in it but it probably doesn't matter for the sake of a prototype. The SDK (or anyone else) would always be free to implement their own aggregation logic and there is no particular benefit to using a runtime implementation other than convenience and some code sharing. Runtime aggregation wouldn't be inherently faster or better.

cijothomas replied:
My question was in context of other .net ecosystem components like Asp.Net Core/Dotnet-Counters tools which want to see aggregates, and would prefer to rely on .NET runtime to provide some built-in aggregators, so they dont need to write one (or depend on otel)

victlu replied:
I think we said .NET will not depend on OTel.  So, there will be some aggregators (that .NET deem important) defined only in .NET.

cijothomas commented on example3/OpenTelemetry.Metric.Sdk/Exporter.cs:
```diff
@@ -0,0 +1,30 @@
+        public abstract void Stop();
+    }
+
+    public class ExportItem
```
is this something which will evolve to match the OTLP protos?

victlu replied:
Just a placeholder to show possible end-to-end.  But, yes, when we're done, this will be evolved or replaced with correct OTLP proto.

cijothomas commented on example3/OpenTelemetry.Metric.Sdk/Aggregators/SumCountMinMax.cs:
```diff
@@ -0,0 +1,42 @@
+            }
+            else
+            {
+                min = Math.Min(min, value);
```
this is just prototype code right and hence not dealing with thread-safety here?

noahfalk replied:
correct, there is at least one todo elsewhere noting we aren't thread-safe

victlu replied:
Agree.  In this particular case, we're safe, as the Collector calls this and that is single instance.

noahfalk replied:
you sure its safe? I thought this code was running synchronously on the thread that called Record() and not inside any lock. This means multiple threads could be executing concurrently in this code block.

cijothomas commented on example3/OpenTelemetry.Metric.Sdk/MetricProvider.cs:
```diff
@@ -0,0 +1,396 @@
+            var token = cts.Token;
+
+            collectTask = Task.Run(async () => {
+                while (!flushCollect && !token.IsCancellationRequested)
```
I read this as: Metrics is "push" based similar to traces..
In the old-but-now-deleted-otel-metric-sdk spec, there was the notion of a Controller, which did the task of pushing the metrics on a schedule, instead of the provider pushing it out to exporters. 
Just wondering if there was any discussion around that model vs push model as in this PR?

victlu committed:
Update from meeting

noahfalk committed:
Replaced required lib+version with MetricSource

MetricSource is optional. Given that this information isn't required in any
other metrics API I am skeptical it has enough value to merit being a mandatory
parameter here. My take is that it should be recommended for library authors
to provide it and optional for app owners.

noahfalk committed:
Updating readme Q&A

victlu committed:
refactoring

cijothomas commented on example3/readme.md:
```diff
@@ -0,0 +1,719 @@
+
+// OR potentially still this if there is a default global factory
+Counter c = new Counter("hats-sold");
+Guage g = new Gauge("temperature");
```
Comparing with tracing:
ActivitySource source = new ActivitySource("name", "ver1.0");
Activity myActivity = source.Create/StartActivity("myactivity", Kind.Client);
myActivity?.SetTag()...
myActivity?.Stop();

The following matches the above model most?
MetricSource source = new MetricSource("name", "ver1.0");
var counter = source.CreateCounter("mycounter"); // if no listener, null returned
counter?.Increment(100, labels);

noahfalk replied:
structurally it is similar, semantically there are some important differences:
a) If a user created the MetricSource and the counter before the SDK registered or before the SDK started listening to a particular counter, returning null at that point means they will never log any measurements on that metric for the lifetime of the app. 
b) Activities are created per-request so there could be 1M activities/sec in a high perf web app. Counters are created once, intended to live for the lifetime of the app, and there should be no counters getting created at steady state. This means there could be lots of allocation caused by steady-state Activity usage but none for steady-state counter usage. Returning null was a perf optimization to compensate for that high allocation rate.

cijothomas replied:
Thanks - this clarifies a lot!

cijothomas replied:
Would you imagine having something like counter.HasListeners() or equivalent? I understand if there are no listeners, these are no-ops, but to help users optimize retrieving some non-trivial label values, a check like this would help. 
Its more of a Otel spec question, as knowing if there is a listener or not alone is not likely sufficient. Probably need more info like which dimensions are doing to be used by the listeners.

noahfalk replied:
> Would you imagine having something like counter.HasListeners() or equivalent?

The prototype has a MeterBase.IsEnabled property that is serving that role at the moment although I don't know if any of our examples bother to call it. As you mentioned the significant savings would be if you can avoid an expensive label computation, otherwise invoking it beforehand saves a few ns to make the call, determine there are no listeners and return.

cijothomas commented on example3/GroceryStoreExample/GroceryStore.cs:
```diff
@@ -0,0 +1,59 @@
+                });
+
+
+            item_counter = new Counter("GroceryStore.item_counter",
```
item_counter = source.CreateCounter("GroceryStore.item_counter",
                new string[] { "Item", "Customer" });
looks easy to write as the user is not requied to pass the source to the Counter's ctor.

The usage of new Ctor(), means the counter will have to return non-null all the time, right? Source.CreateCounter() allows us to return null if no listener expressed interest in this counter. (This is similar to ActivitySource.CreateActivity() as well)

cijothomas replied:
The OpenTelemetry.Metric.API project allows MeterProvider.Global.GetMeter(name,ver).CreateCounter(countername) though.

noahfalk replied:
> The usage of new Ctor(), means the counter will have to return non-null all the time, right?

Yep! I doubt we'd want to return null even if we had the option to do so with a factory. Even if no listener is interested right away, they might be interested later. And if we don't create the counter now then there will no way to send data later.

cijothomas commented on example3/Microsoft.Diagnostics.Metric/AggregationConfiguration.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Microsoft.Diagnostics.Metric
+{
+    public static class AggregationConfigurations
```
Would .NET want to expose this? or just expose listener , and its upto listeners (Otel etc) to build aggregators..

noahfalk replied:
I think its at least on the table that .NET exposes stuff like this. In this case these types aren't implementations of aggregators, only descriptions of what aggregation the instrumenting dev recommends to occur. I envision use-cases where a library developer wants to communicate things like "I recommend  tracking these times as percentiles reporting the 50th, 95th, and 99th percentile and bucket requests by response code". A listener isn't obligated to do what the instrumenter asks, but regardless the instrumenter still needs some API to describe their recommendation.

cijothomas replied:
thanks for clarifying. yes this makes sense.

victlu replied:
This should be part of the "hinting" rather than specific to Aggregation.  @noahfalk, I though this was more for the .NET specific aggregator in the "SDK" portion (.NET runtime consumption of Listener rather than OTel) of .NET API?

noahfalk replied:
> This should be part of the "hinting" rather than specific to Aggregation.

Other than aggregation hints are there other hints you have in mind?

> I though this was more for the .NET specific aggregator in the "SDK" portion (.NET runtime consumption of Listener rather than OTel) of .NET API?

I'm imagining it being applicable for both, though all listeners are free to ignore it. I'm assuming this is something OTel is going to want but they haven't gone too far down the path yet. For example the only real difference between a Counter and a Gauge is that one hints the numbers should be aggregated by adding them and the other should be aggregated by overwriting the last value. Effectively it is a hint embedded in the choice of instrument type. I am guessing that there are too many useful possible hints to create a separate instrument for each one so at some point OTel will define instrumentation side hinting more explicitly.

victlu replied:
Another potentially important hint will be Units. Maybe Monotonic is a hint as well. Maybe Min/Max limits/thresholds, etc...
 
Bucket configuration probably fits with Histogram aggregation, maybe.

Maybe we have multiple aggregators at the same time.  I.e. We want Sum, Histogram, and Summary together.

cijothomas commented on example3/Microsoft.Diagnostics.Metric/MetricSource.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Microsoft.Diagnostics.Metric
+{
+    public class MetricSource
```
how does MeterListener gets attached to a MeterSource? In tracing, there is the static method ActivitySource.AddListener(listener)

noahfalk replied:
In the current code you call the MetricListener.Start() function. I hadn't paid too much attention to where it should show up in the API.

noahfalk replied:
I'll admit I am liking Start() better than the static MetricSource.AddListener() API at the moment. Maybe if the mismatch is bugging people we can add ActivityListener.Start() to restore parity : )

cijothomas replied:
Otel end users won't care as we'll hide it under.
Start() logically makes sense - as it indicates starting of the listening. I'd still prefer consistency :-)

victlu committed:
Add more discussion items

victlu committed:
Benchmark for ScalarDataPoint

victlu committed:
Add Benchmark for how we pass labels

victlu committed:
Add additional Bench for passing labels

victlu committed:
update results for Labelpassing bench

victlu committed:
Update result from LabelPassing bench

noahfalk committed:
Aligning on OTel's current naming

Regardless whether we stick with this naming I think it helps
keep the mapping clear to match up the names.

noahfalk committed:
WIP

noahfalk committed:
tests passing and perf benchmarks recorded

noahfalk committed:
Wrap up the dynamic label tuple experiment for now

Seemed successful : )

I don't think I supported other instruments besides Counter
but not expecting it to be significantly different. Also I
probably broke the BoundInstruments, not predicting its hard
to fix.

noahfalk committed:
Update readme.md

reyang commented on example3/readme.md:
```diff
@@ -659,6 +659,9 @@ Given pass experiences, handling and management of LabelSet will be a hot topic
 
+## How should we manage the lifetime of instruments/meters?
+
+We probably want to clean them up as some point so we should figure out when that happens. Does Meter or Instrument implement IDisposable?
```
I think we need to have IDisposable.
I think the instrument lifecycle is attached to meter - if meter gets disposed, all the instruments "under" that meter should also be disposed. Instrument can also be disposed individually.

noahfalk replied:
What if we said Meter is the only thing that is IDisposable and if you want counters to have different lifetimes they should be in a different Meter, does that feel too restrictive?

reyang replied:
I think it is perfectly fine.
Curious - is there a reason you would prefer not to have IDisposable on Instruments (Counter, Gauge, ...)?

noahfalk replied:
The usual expectation is that any type implementing IDisposable needs to be disposed. I was hoping to avoid making users think they need to call Dispose() on every instrument individually when disposing the Meter alone would work just as well.

noahfalk committed:
Fix variable naming

noahfalk committed:
Refactor MeterInstrument -> UnboundMeterInstrument

noahfalk committed:
MeterInstrumentBase -> MeterInstrument

noahfalk committed:
Eliminate bound instruments

noahfalk committed:
Create instrument via Meter factory

noahfalk committed:
Add IDisposable on Meter and listener

noahfalk committed:
switch from listener delegates to virtual methods

noahfalk committed:
Readme updates

victlu commented:
Question: OTel adopted attributes with name+type.  Will .NET API also switch to attributes, or will it continue to take value of type string?

noahfalk commented:
> Question: OTel adopted attributes with name+type. Will .NET API also switch to attributes, or will it continue to take value of type string?

For naming I'm happy to use "Attribute" rather than "Label." For the data types I still expect it to be challenging to make effective use of non-string label types so I was not planning to do so. The discussion [here](https://github.com/noahfalk/opentelemetry-dotnet/tree/design_experiment/example3#should-we-standardize-that-all-dimension-values-are-strings-in-the-api-or-do-we-need-to-allow-a-broader-set-of-types) looks up-to-date but if you have concerns or think we should be doing something else I am glad to discuss any time.

victlu commented:
It was a clarifying question.  Given recent SIG discussions leaning toward Attributes, I was just wondering if that decision would sway you. My preference is also keeping it as strings as well.

noahfalk committed:
Update readme.md

noahfalk committed:
Removed static labels

noahfalk committed:
Rename ObservableCounter -> CounterFunc

victlu committed:
Interpretation of current API spec

victlu committed:
Add example implementation of current API spec

victlu committed:
Fixup stuff

noahfalk committed:
Cleaner implementation of AggregationStore

noahfalk committed:
Fix broken HappyPath test

Multiple tests used Provider.GetMeter("test") and then created a
counter called "counter". Because Provider is a static and
tests run in parallel both were sharing the same Meter object and
attempting to create the same counter which threw an Exception.

noahfalk committed:
Eliminated the hint API (for now)

We plan to have a hinting API again in the future so I expect this to
return in some form. I am just aligning the API to the current spec.

victlu committed:
Refactor Api2

victlu committed:
Refactoring...

victlu committed:
cleanup

victlu committed:
Add IDisposable

noahfalk committed:
Add SDK instrument configuration

noahfalk committed:
Get rid of the 5X in the benchmarks

noahfalk committed:
Add a simple Prometheus exporter

noahfalk committed:
Add Distribution instrument

I don't know what the spec will name it, but I assume we are
going to have one. This also implements a simple ExponentialHistogram
in the SDK to back it.

victlu committed:
dotnet-format fixes

victlu committed:
misspell

victlu committed:
wrapping

victlu committed:
Merge branch 'main' into design_experiment

victlu committed:
dotnet-format

victlu committed:
remove old code

victlu committed:
allow CRLF for .NET/C# files

victlu committed:
Change to LF

victlu committed:
Get example working again

victlu committed:
Hook up Api2 to our sdk

victlu committed:
Add notification to API2

victlu committed:
Add <T> to all instruments

noahfalk committed:
Removing the tagging interfaces

The API review folks generally don't like to have interfaces with no
members and we can do a type check that avoids needing them. If we
want a better non-type based check I think the best approach will be
to decide on a hinting API.

noahfalk committed:
Fix perf regressions

The perf regression from adding the T was larger than I remembered,
about 15ns. I tracked it down to perf impact of making generic virtual
calls. Unlike normal virtual calls that can be resolved by looking up
the function pointer in a known slot in a table, the generic T requires
a dictionary lookup in the runtime. Thankfully we can work around
that by providing type specific overloads.

victlu committed:
Target multiple frameworks

victlu committed:
Instroduce Metric API into OpenTelemetry

victlu committed:
Hook up MDM to see how it feels.

noahfalk committed:
Use object type for label values

After continued reflection I've reversed my earlier position on using
object type for label values. I think most others wanted object and were
just defering to me as the holdout so probably this isn't contentious?

Also fixed a bug in the handling of 3 labeled instruments
(hardcoded 2 at MeterInstrument.cs:165 should have been a 3)

noahfalk committed:
Get rid of UnboundMeterInstrument

noahfalk committed:
Remove MeasurementObserver and create ObservableMeterInstrument base class

noahfalk committed:
Add description and unit

noahfalk committed:
Eliminate unneeded protected members

noahfalk committed:
Eliminate build warnings

noahfalk committed:
Update MeterInstrumentListener API

victlu committed:
Add Bench to compare cost of API/SDK vs IFx

victlu committed:
Merge branch 'design_experiment' of https://github.com/noahfalk/opentelemetry-dotnet into design_experiment

victlu committed:
Update bench results

victlu committed:
Fix bad checkin

victlu committed:
Remove strong name signing

victlu committed:
Remove reference

victlu committed:
Make it build!

victlu committed:
Update to new Listener hookup

victlu committed:
Use IFxExtension to handle multi-dim labels

victlu committed:
Setup Example to send to GenevaAgent

victlu committed:
Try out passing KVP in bench

victlu committed:
Not sure why CreateReadOnlySpan doesn't work!

victlu committed:
Get bench working for CreateReadOnlySpan

victlu committed:
Bench for CreateReadOnlySpan of 5 KVP
